File: js/perf-policy.js
import { state } from './state.js';
import { gpuThrottleReadbackMul } from './gpu-throttle.js';

const MILESTONE_EVENT_KINDS = new Set([
  'appeared',
  'born',
  'mated',
  'gaveBirth',
  'hunted',
  'preyedOn',
  'died',
  'stage',
]);

export function effectiveSnapshotIntervalSec()
{
  return Math.max(0.5, state.snapshotIntervalSec || 10);
}

export function effectiveHeartbeatIntervalSec()
{
  const base = Math.max(1, state.heartbeatIntervalSec || 5);
  const speed = Math.max(1, state.speed || 1);
  if (speed < 4) return base;
  return base * Math.max(1, speed / 3);
}

export function shouldRunBehaviorThisSubstep(substep, substepCount)
{
  const speed = Math.max(1, state.speed || 1);
  if (speed < 5) return true;
  if (!Number.isFinite(substep) || !Number.isFinite(substepCount) || substepCount <= 0) return true;
  return substep >= substepCount - 1;
}

export function effectiveReadbackEveryMs(baseMs)
{
  if (state.batchMode) return 0;
  const speed = Math.max(1, state.speed || 1);
  const tier = state.gpuTelemetry?.qualityTier ?? 0;
  const tierMul = tier >= 3 ? 1.8 : tier >= 2 ? 1.35 : 1;
  const throttleMul = gpuThrottleReadbackMul();
  return baseMs * Math.max(1, speed / 3) * tierMul * throttleMul;
}

export function effectiveScrubTickRefreshMs()
{
  return 800;
}

export function shouldPersistCreatureEvent(kind, creatureId = null)
{
  const speed = Math.max(1, state.speed || 1);
  if (speed < 5) return true;
  if (MILESTONE_EVENT_KINDS.has(kind)) return true;
  if (
    creatureId != null
    && state.selected
    && !state.selected.dead
    && state.selected.id === creatureId
  )
  {
    return true;
  }
  return false;
}

export function timelineWritePressure()
{
  const speed = Math.max(1, state.speed || 1);
  if (speed >= 8) return 'high';
  if (speed >= 5) return 'medium';
  return 'low';
}

File: godot/WildlandsEcoSim/EcoSim.Core/Sim/PerfPolicy.cs
namespace EcoSim.Core.Sim;

public static class PerfPolicy
{
    private static readonly HashSet<string> MilestoneKinds = new(StringComparer.Ordinal)
    {
        "appeared", "born", "mated", "gaveBirth", "hunted", "preyedOn", "died", "stage",
    };

    public static double EffectiveSnapshotIntervalSec(double snapshotIntervalSec) =>
        Math.Max(0.5, snapshotIntervalSec);

    public static double EffectiveHeartbeatIntervalSec(double heartbeatIntervalSec, double speed)
    {
        double baseSec = Math.Max(1, heartbeatIntervalSec);
        speed = Math.Max(1, speed);
        if (speed < 4) return baseSec;
        return baseSec * Math.Max(1, speed / 3);
    }

    public static bool ShouldRunBehaviorThisSubstep(int substep, int substepCount, double speed)
    {
        speed = Math.Max(1, speed);
        if (speed < 5) return true;
        if (substepCount <= 0) return true;
        return substep >= substepCount - 1;
    }

    public static bool ShouldPersistCreatureEvent(string kind, int? creatureId, int? selectedId, bool selectedDead)
    {
        return true;
    }

    public static bool ShouldPersistCreatureEventAtSpeed(string kind, double speed, int? creatureId, int? selectedId, bool selectedDead)
    {
        speed = Math.Max(1, speed);
        if (speed < 5) return true;
        if (MilestoneKinds.Contains(kind)) return true;
        if (creatureId != null && selectedId == creatureId && !selectedDead) return true;
        return false;
    }

    public static string TimelineWritePressure(double speed)
    {
        speed = Math.Max(1, speed);
        if (speed >= 8) return "high";
        if (speed >= 5) return "medium";
        return "low";
    }

    public static double EffectiveScrubTickRefreshMs() => 800;
}
